src/DocumentationViewer.js

```diff
@@ -0,0 +1,219 @@
+    }
+  });
+
+  search_entry.connect("search-changed", () => {
+    listbox.invalidate_filter();
+  });
+
+  search_bar.bind_property(
+    "search-mode-enabled",
+    search_button,
+    "active",
+    GObject.BindingFlags.BIDIRECTIONAL,
+  );
+
+  search_button.connect("toggled", () => {
+    if (search_button.active) sidebar_button.active = true;
+  });
+
+  listbox.set_sort_func(sort);
+  listbox.invalidate_sort();
+  listbox.set_filter_func(filter);
+
+  getDocs(base_path)
+    .then((docs) => {
+      for (const doc in docs) {
+        const row = new Adw.ActionRow({
+          title: doc,
+        });
+        row.add_suffix(new Gtk.Image({ icon_name: "go-next-symbolic" }));
+        listbox.append(row);
+      }
+      listbox.connect("row-selected", (_, row) => {
+        webview.load_uri(docs[row.title]);
+      });
+      return docs;
+    })
+    .catch(logError)
+
+
+
+  function sort(row1, row2) {
+    return row1.title > row2.title;
+  }
+
+  function filter(row) {
+    const re = new RegExp(search_entry.text, "i");
+    return re.test(row.title);
+  }
+
+  const action_documentation = new Gio.SimpleAction({
+    name: "documentation",
+    parameter_type: null,
+  });
+  action_documentation.connect("activate", () => {
+    window.present();
+  });
+  application.add_action(action_documentation);
+  application.set_accels_for_action("app.documentation", ["<Control><Shift>D"]);
+}
+
+async function getDocs(base_path) {
+  try {
+    const docs = {};
+    const dirs = await list(base_path);
+
+    for (const dir of dirs) {
+      try {
+        const index = await readDocIndex(base_path, dir);
+        const namespace = `${index["meta"]["ns"]}-${index["meta"]["version"]}`;
+        const path = `file://${base_path}/${dir}/index.html`;
+        docs[namespace] = path;
+      } catch (e) {
+        // Ignore the error if the dir does not contain index.json
+        if (!e.matches(Gio.IOErrorEnum, Gio.IOErrorEnum.NOT_FOUND)) logError(e);
+      }
+    }
+
+    return docs;
+  } catch (e) {
+    logError(e);
+  }
+}
+
+async function readDocIndex(base_path, dir) {
+  const file = Gio.File.new_for_path(`${base_path}/${dir}/index.json`);
+  const [json, _] = await file.load_contents_async(null);
+  return JSON.parse(new TextDecoder().decode(json));
+}
+
+async function list(dir_path) {
+  // List all files in dir_path
+  try {
+    const dir = Gio.File.new_for_path(dir_path);
+    const files = [];
+    const enumerator = await dir.enumerate_children_async(
+      "standard::name",
+      Gio.FileQueryInfoFlags.NOFOLLOW_SYMLINKS,
+      GLib.PRIORITY_DEFAULT,
+      null,
+    );
+    while (true) {
+      const infos = await enumerator.next_files_async(
+        10,
+        GLib.PRIORITY_DEFAULT,
+        null,
+      );
+      if (infos.length === 0) break;
+      for (const info of infos) {
+        files.push(info.get_name());
+      }
+    }
+    return files;
+  } catch (e) {
+    logError(e);
+  }
```

Comment:
since you have a error handler up the chain in https://github.com/sonnyp/Workbench/pull/358/files#diff-6533bf97682b99bf434791a454793569a69668f437288ddcedac958454abf4a7R122, you can remove the try/catch in this function